tensorflow/lite/kernels/activations.cc

```diff
@@ -1036,64 +1051,56 @@ TfLiteStatus SigmoidEval(TfLiteContext* context, TfLiteNode* node) {
-                         GetTensorData<float>(output),
-                         CpuBackendContext::GetFromContext(context));
+  if (kernel_type == kReference) {
+    reference_ops::Softmax(op_params, GetTensorShape(input),
+                           GetTensorData<float>(input), GetTensorShape(output),
+                           GetTensorData<float>(output));
+  } else {
+    optimized_ops::Softmax(op_params, GetTensorShape(input),
+                           GetTensorData<float>(input), GetTensorShape(output),
+                           GetTensorData<float>(output),
+                           CpuBackendContext::GetFromContext(context));
+  }
   return kTfLiteOk;
 }
 
 template <typename In, typename Out>
 TfLiteStatus SoftmaxQuantized(TfLiteContext* context, const TfLiteTensor* input,
-                              TfLiteTensor* output, SoftmaxOpData* data) {
-  optimized_ops::Softmax(data->params, GetTensorShape(input),
-                         GetTensorData<In>(input), GetTensorShape(output),
-                         GetTensorData<Out>(output));
-  return kTfLiteOk;
-}
-
-template <>
-TfLiteStatus SoftmaxQuantized<int8_t, int8_t>(TfLiteContext* context,
-                                              const TfLiteTensor* input,
-                                              TfLiteTensor* output,
-                                              SoftmaxOpData* data) {
-#ifdef TFLITE_SOFTMAX_USE_UINT16_LUT
-  optimized_ops::SoftmaxInt8LUT(
-      data->params, GetTensorShape(input), GetTensorData<int8_t>(input),
-      GetTensorShape(output), GetTensorData<int8_t>(output));
-#else
-  optimized_ops::Softmax(data->params, GetTensorShape(input),
-                         GetTensorData<int8_t>(input), GetTensorShape(output),
-                         GetTensorData<int8_t>(output));
-#endif
-  return kTfLiteOk;
-}
-
-template <>
-TfLiteStatus SoftmaxQuantized<uint8_t, uint8_t>(TfLiteContext* context,
-                                                const TfLiteTensor* input,
-                                                TfLiteTensor* output,
-                                                SoftmaxOpData* data) {
+                              TfLiteTensor* output, SoftmaxOpData* data,
+                              KernelType kernel_type) {
+  if (kernel_type == kReference) {
+    reference_ops::Softmax(data->params, GetTensorShape(input),
+                           GetTensorData<In>(input), GetTensorShape(output),
+                           GetTensorData<Out>(output));
+  }
 #ifdef TFLITE_SOFTMAX_USE_UINT16_LUT
-  optimized_ops::SoftmaxInt8LUT(
-      data->params, GetTensorShape(input), GetTensorData<uint8_t>(input),
-      GetTensorShape(output), GetTensorData<uint8_t>(output));
-#else
-  optimized_ops::Softmax(data->params, GetTensorShape(input),
-                         GetTensorData<uint8_t>(input), GetTensorShape(output),
-                         GetTensorData<uint8_t>(output));
+  else if ((std::is_same<In, uint8_t>::value &&
```

Comment:
This is not a good style. Please do not mix the "else" condition and #ifdef macro.

```
#ifdef TFLITE_SOFTMAX_USE_UINT16_LUT
if ((std::is_same<In, uint8_t>::value && ...
```

Reply:
I found it more readable like that but I made the change. I restored the functions as they were and just added the reference case to them.